### ISSUE TITLE
Feature or Assistance: Get compiled validation code

### ISSUE THREAD

**Issue by tamusjroyce** [ajv-validator/ajv#1329]
**What version of Ajv you are you using?** "ajv": "6.12.6"

**What problem do you want to solve?** Either documentation or a new method to easily get a single compiled validation method as text

**What do you think is the correct solution to problem?** Better documentation, a new method to get compiled source-code, or answer that 7 beta has this working

**Will you be able to implement it?** I see the method when stepping into node_modules/ajv. But I haven't been able to pin-point why source code is a wrapper around the validation function I wish to see

`npm install quicktype-core json-schema-to-openapi-schema avj && node index.js`

```
// Typescript index.js, Goal is to infer schema from data and generate appropriate compiled validation
//    Feeds documentation, used in Azure Trigger which cannot include external dependencies,
//    feeds application validation, samples existing endpoint data and gives UI dev-tooling for validation, etc...

import {
    quicktype,
    InputData,
    jsonInputForTargetLanguage
} from "quicktype-core";

import * as toOpenApi from 'json-schema-to-openapi-schema';
import * as Ajv from 'ajv';

async function quicktypeJSONSchema(jsonSchemaStrings: string[], typeName, targetLanguage = 'JSON Schema') {
    const schemaInput = jsonInputForTargetLanguage(targetLanguage);
  
    // We could add multiple schemas for multiple types,
    // but here we're just making one type from JSON schema.
    await schemaInput.addSource({ name: typeName, samples: jsonSchemaStrings });
  
    const inputData = new InputData();
    inputData.addInput(schemaInput);
  
    const result = await quicktype({
      inputData,
      lang: targetLanguage,
    });

    if (targetLanguage === 'JSON Schema') {
        return JSON.parse(result.lines.reduce((str, line) => str + line, ''));
    }

    return result.lines.reduce((str, line) => str + line + "\r\n", '');
}

const samples = [{
  "id": 28,
  "Title": "Sweden",
  "test" {
    "a": 1,
    "option_a": "a"
  },
  "required": "yes"
}, {
  "id": 56,
  "Title": "USA",
  "Optional": 5,
  "test" {
    "a": 2,
    "option_b": "b"
  },
  "required": "ok"
}, {
  "id": 89,
  "Title": "England",
  "test" {
    "a": 3,
    "option_c": "c"
  },
  "required": "nice"
}];

const generateSchema = async function(data): Promise<any> {
    let responseMessage = [];
    var ajv = new Ajv(); // options can be passed, e.g. {allErrors: true}
    ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-06.json'));
    ajv.addFormat('integer', new RegExp('(?<=\s|^)\d+(?=\s|$)'));

    try {
        const jsonSchema = data; // Code this is based on grabs data from database
        const openApi = toOpenApi(jsonSchema);
        const typescriptSchemaCode = await quicktypeJSONSchema(resources.map(doc => JSON.stringify(doc)), collection, 'TypeScript');
        const csharpSchemaCode = await quicktypeJSONSchema(resources.map(doc => JSON.stringify(doc)), collection, 'C#');

        ajv._opts.allErrors = false; // This collects all errors instead of failing validation fast and returning only the first error

        // WARNING: validationCode is incorrect...but it is obtainable. Out of time for this spike!
        // TODO: Get the real source code for validation by stepping into avj.compile...refVal should not be needed!!! Unexpected source code is returned
        ajv._opts.sourceCode = true; // Our purpose is to get the source code
        const code = ajv.compile(jsonSchema).source['code'] || 'validate = function(data) { if (data) { this.errors = [{ errorMessage: "data should not be defined" }]; return false; } return true; }';
        const validationCode = '(data) => { const refVal = []; ' + code.split('return validate;').join('') + '  const isValid = validate(data); return { isValid, errors: validate.errors }; }';

        // TODO: Why is validationCode showing this? (v + '') inside the call-stack shows source-code correctly

        responseMessage.push({ jsonSchema, openApi, typescriptSchemaCode, csharpSchemaCode, validationCode });
    } catch(ex) {
        console.log(`bad resource ${data} not found`, ex);
    }

    return responseMessage;
};

console.log(generateSchema(samples));
```

**Comment by epoberezkin** [comment]
With ajv v6 you can get standalone code for a single validation function using ajv-pack package (or ajv-cli that uses it).

It does not support all possible schemas - there are limitations - see the docs of ajv-pack.

V7 will have much a better support for it, land you will be able to generate a module with multiple validation functions (which is the only way to do it in case of mutual recursion), but it is not there yet. I  am currently working on it, it will be done before v7 becomes the main version.

**Comment by epoberezkin** [comment]
> I haven't been able to pin-point why source code is a wrapper around the validation function I wish to see

If I understood the question correctly, that is the only way to construct closure without holding the whole global space in the closure. With standalone code the same values will be module scoped so the closure won’t be needed.

**Comment by tamusjroyce** [comment]
Thank you @epoberezkin 

I am not wanting global space, but inline or the entire block of code.

`function validate15(data, {dataPath="", parentData, parentDataProperty, rootData=data}={}){
let vErrors = null;
let errors = 0;
if(!(validate16(data, {dataPath,parentData,parentDataProperty,rootData}))){
vErrors = vErrors === null ? validate16.errors : vErrors.concat(validate16.errors);
errors = vErrors.length;
}
validate15.errors = vErrors;
return errors === 0;
}`

This is with 7-beta. Why is there a validate16 within validate15 without validate16 also being returned? 

Awesome to hear this might be an up-and-coming feature. Thank you for all the work and this awesome library. If you have a ticket open for getting the entire validation code in one call, please close this.

Hope this is helpful for anyone who wants to infer schema from data and compile stand-alone validation

**Comment by tamusjroyce** [comment]
Closing with https://github.com/ajv-validator/ajv/pull/1332/commits (if there are issues, I will open another ticket or hopefully include a PR)

**Comment by epoberezkin** [comment]
It is now released as v7.0.0-beta.8 - it can generate a standalone module code - see standalone.md doc (CLI is not done yet but you can try using it programmatically, it works and passes all the tests).

To your specific question, validate16 is another function that is in the closure of validate15. If you use standalone functions they both will be in the generated module.

This approach avoids duplication of shared logic - validate16 can be used in multiple functions, that is why it is not returned together

**Comment by tamusjroyce** [comment]
Just an update. But I was able to use quicktype / quicktype-core to generate compiled validation when converting json schema to typescript interfaces. But I really like the standalone for ajv. I plan on comparing if there are any differences. But ajv is best runtime.
